docetl/dataset.py

```diff
@@ -0,0 +1,326 @@
+
+    def load(self) -> List[Dict]:
+        """
+        Load the dataset from the specified path or return the in-memory data.
+
+        Returns:
+            List[Dict]: A list of dictionaries representing the dataset.
+
+        Raises:
+            ValueError: If the file extension is unsupported.
+        """
+        if self.type == "memory":
+            return self._apply_parsing_tools(self.path_or_data)
+
+        _, ext = os.path.splitext(self.path_or_data.lower())
+
+        if ext == ".json":
+            import json
+
+            with open(self.path_or_data, "r") as f:
+                data = json.load(f)
+        elif ext == ".csv":
+            import csv
+
+            with open(self.path_or_data, "r") as f:
+                reader = csv.DictReader(f)
+                data = list(reader)
+        else:
+            raise ValueError(f"Unsupported file extension: {ext}")
+
+        return self._apply_parsing_tools(data)
+
+    def _apply_parsing_tools(self, data: List[Dict]) -> List[Dict]:
+        """
+        Apply parsing tools to the data.
+
+        Args:
+            data (List[Dict]): The data to apply parsing tools to.
+
+        Returns:
+            List[Dict]: The data with parsing tools applied.
+
+        Raises:
+            ValueError: If a parsing tool is not found or if an input key is missing from an item.
+        """
+        for tool in self.parsing:
+            input_key = tool["input_key"]
+            if tool["function"] in PARSING_TOOLS:
+                func = PARSING_TOOLS[tool["function"]]
+            elif (
+                self.user_defined_parsing_tool_map
+                and tool["function"] in self.user_defined_parsing_tool_map
+            ):
+                func = eval(
+                    self.user_defined_parsing_tool_map[tool["function"]].function_code
+                )
+            else:
+                raise ValueError(
+                    f"Parsing tool {tool['function']} not found. Please define it or use one of our existing parsing tools: {PARSING_TOOLS.keys()}"
+                )
+
+            output_key = tool["output_key"]
+            function_kwargs = tool.get("function_kwargs", {})
+            new_data = []
+
+            def process_item(item):
+                if input_key not in item:
+                    raise ValueError(f"Input key {input_key} not found in item: {item}")
+                result = func(item[input_key], **function_kwargs)
+                if isinstance(result, list):
+                    return [item.copy() | {output_key: res} for res in result]
+                else:
+                    return [item | {output_key: result}]
+
+            with ThreadPoolExecutor() as executor:
+                futures = [executor.submit(process_item, item) for item in data]
+                new_data = []
+                for future in as_completed(futures):
+                    new_data.extend(future.result())
+
+            data = new_data
```

Comment:
is the intent that each parsing (tool) round overrides the last round's computed results? this seems to be what's happening.

Reply:
oh! this does make sense. At first, I didn't get the fact that parsing is meant to compose the parsing tools as a pipeline.

Reply:
yep :-) though in practice i don't expect there to be more than one tool lol